### DISCUSSION TITLE
Kiali Full Traffic Visualization (Start To End)

### DISCUSSION THREAD

**Discussion by hassanbsee2071** [kiali/kiali#7892] [Q&A]
### What do you want to discuss?

When I see the visualization in kiali traffic graph, I could not see the full visualization of traffic flow, for example in below pictures, it is calling **shared-miscellaneous-gateway-deployment**. But after that it is not showing where traffic is going, I have to right click on the workload and then have to select **Details**. Is there any option in kiali by which we can see the full visualization from start to end? 
![image (3)](https://github.com/user-attachments/assets/253aedff-1f8f-411d-a3be-870b6d10b2b3)
![image (4)](https://github.com/user-attachments/assets/28122af7-2a49-415f-814c-45d0fe5e1e90)
![image (5)](https://github.com/user-attachments/assets/dd454bb7-ee10-41ab-80ef-ec7e156ca36e)
![image (6)](https://github.com/user-attachments/assets/c00df7b7-a692-44af-9396-85e31ea8512a)

**Comment by jmazzitelli** [comment]
_(converted this to Q&A Discussion as this is not an "issue")_

**Comment by jmazzitelli** [comment]
In the Namespace dropdown menu, did you select all the namespaces involved in the traffic you are interested in?

Look at the icons in the nodes on the right side of your graph
![image](https://github.com/user-attachments/assets/c9130f97-6f8a-4264-8609-8a71c5b87cce)
That little "network" icon indicates that's going to another namespace that you have not selected in the namespace dropdown.

Go to the Namespace drop down and select all the namespaces involved in the traffic you are interested in and it should show you everything.

@jshaughn can confirm the above - I'm pretty sure that's what that icon means.

**Reply by hassanbsee2071** [reply]
Thanks for the reply. In below screen shot as you can see, I have selected 2 namespaces, but result is same.

**How I am doing it**

First I select **Applications**, then filter the application, click on it and then goes to **Show node graph**. In the namespace selection I have already selected two namespaces, flights and shared. Kiali is showing that it is going from flights to shared namespace for **shared-miscellaneous-gateway-deployment**, but from that it goes to **shared-miscellaneous-core** in same namespace but kiali is not showing it

![image](https://github.com/user-attachments/assets/f8c16282-8d97-4645-ba93-ffa7fcbefce8)

![image](https://github.com/user-attachments/assets/6b402aca-906d-4f6f-8887-57147d9facff)

![image](https://github.com/user-attachments/assets/29763086-8918-456e-95a7-570110fa00a4)

**Comment by jmazzitelli** [comment]
The node graph is different from the main graph. The node graph will only show one level going in and one going out of the node selected (that's why it's called the "node" graph because its concentrating on just the node in question - it will only show things directly going into and out of the selected node).

Go to the main traffic graph page (don't filter on any applications), select all the namespaces, and look at the graph there. See what it looks like.

**Reply by hassanbsee2071** [reply]
But how can i filter the application there? We have more than 1000 of deployments. That graph is totally not understandable?

**Reply by hassanbsee2071** [reply]
There is option called Traffic Graph, which shows all the services. But we have more than 1000 of applications and that graph looks so messy like a web. Not able to filter specific application? Is there any way to look for specific application in Traffic Graph

**Reply by jmazzitelli** [reply]
That's where the Hide/Find features can help with.

Here's a blog when this was introduced with some docs on this feature: https://medium.com/kialiproject/kiali-releases-1-34-to-1-39-overview-587f33fac41a#3962

Here is an old video of a very old version of Kiali, but the feature is the same today: https://youtu.be/Yl8gQffBGNs?feature=shared

**Comment by jmazzitelli** [comment]
@jshaughn do we have better find/hide docs anywhere on kiali.io? I couldn't find anything other than a sentence here or there on it. We should have a dedicated page or section somewhere on the find/hide feature (what are the more useful expressions would be helpful at least in an FAQ).

**Reply by jshaughn** [reply]
The best Hide/Find doc is in Kiali itself, click the info icon next to the Hide/Find input boxes and it explains the DSL (the supported expression language) in a lot of detail.  Note that hitting `Tab` will autocomplete expressions, although not values.

If you have a lot of services and traffic connections I think it's normal to expect a complicated graph.  What is important to know is that you have a variety of options to simplify the graph, even on your initial visit.  The graph options are all bookmark-able, so If you have preferred views, you can bookmark them, and use them to visit Kiali.  Also, in the Kiali CR you can pre-define graph-hide expressions, and even have them apply by default (see spec.kiali_feature_flags.ui_defaults.graph).

**Reply by jshaughn** [reply]
You can also play with many things, The Traffic dropdown allows you to filter by protocol, boxing options, service node injection, and much more can be set in the Display dropdown, several layouts are available via the on-graph icons, as well as accelerators for hiding edges (like hide all healthy edges).

**Reply by jshaughn** [reply]
Going back to your original comment, the main graph has a lot of hover/selection highlighting to identify flows.  So, given a dense graph, you could do something like a graph find of `app=<myAppName>` and then when you see the node, select it.  This will dim all paths not including that node, as well as give you some detail in the side panel.

**Reply by hassanbsee2071** [reply]
Thanks a lot for your reply. Really appreciate. I have label on deployment. 
app=flights-api-node. Can we filter on the basis of deployment label?

**Reply by jshaughn** [reply]
yes, we do support hide/find on labels, see the help

**Reply by jshaughn** [reply]
![image](https://github.com/user-attachments/assets/654bfcae-2687-4028-affb-77d2fdb8f16f)

Some examples:

![image](https://github.com/user-attachments/assets/304baf10-d4c1-46da-9910-3bbbffb86130)

**Reply by hassanbsee2071** [reply]
I tried finding the application by labels, but still cant see anything useful. This is with only two namespaces, this app is communicating with other namespaces as well. As you can see in the picture, after adding label, yellow small box appears, I click on this line but still it is messy. How can I hide all other lines and only displays this

<img width="1727" alt="image" src="https://github.com/user-attachments/assets/74a68c20-9721-499a-bc03-a213ee59e6e6">

**Reply by jmazzitelli** [reply]
> How can I **_hide_** all other lines and only displays this

Use the "hide" feature to hide things :) 

The "show" feature highlights items in the graph - the "hide" feature makes them disappear.

**Reply by hassanbsee2071** [reply]
I am sorry if asking dumb questions, I understand the hide feature will hide :-)
I have 100 deployments in flights namespace, and I want to see only app=flights-api-v3 , so I added find: app=flights-api-v3 and to hide all others, hide: **name not contains flights-api-v3,** also tried with **app not contains flights-api-v3**, so It only displays api-v3, but graph becomes empty, let me play for some more time. 

<img width="1719" alt="image" src="https://github.com/user-attachments/assets/ae409b8f-9ace-4352-8640-5a896d1f9c00">

**Reply by jshaughn** [reply]
Not sure if it's an issue but you may want to avoid the "name" operand in the hide expression. Notes: 
* tests against canonical service, operation, service and workload names
* The "name" operand expands internally to an "OR" expression (an "AND" when negated).

**Reply by jshaughn** [reply]
I think the thing you really want to do may be at odds with what we offer.  I think you want to be able to limit the graph to all paths that include a particular node.  The closest way to do this is what I mentioned above, to select the node (click on it), which will dim all paths that it is not involved in.  Then you could zoom to examine the path more closely.  As mentioned above, double-clicking into the node graph will limit the graph to the immediate incoming and outgoing traffic.  I actually think it could be a nice feature to do what you really want, and have an option to immediately Hide nodes not connected to the selected node.

**Reply by jmazzitelli** [reply]
> I actually think it could be a nice feature to do what you really want, and have an option to immediately Hide nodes not connected to the selected node.

Which is similiar to the dimming feature where all edges/nodes are dimmed that are not related to selected node. We could, as you say, add some configuration option (in the Display drop-down perhaps) to have this option:

- Emphasis style
[ ] dim unselected
[ ] hide unselected

I can't figure out good names for those options but you get the idea - either have the graph "dim" or "hide" those nodes/edges that are not selected. The default would be "dim" since that's how it works today.

We'd have to see how that works - because I could see the graph "jumping" around as nodes are hidden, so it could get confusing and might hurt your eyes. :)

**Reply by hassanbsee2071** [reply]
@jshaughn and @jmazzitelli , thank you so much for your prompt responses and @jmazzitelli , I agree with you :-). See below picture, It dims the other line, but still very hard to figure out the actual flow. I only want to show the highlighted line and hide all other lines so it be visualized.

<img width="1718" alt="image" src="https://github.com/user-attachments/assets/16b8d257-a1a2-400e-8b45-ece4dddbf0a6">

**Reply by jmazzitelli** [reply]
Another idea: maybe in the "Hide" filter, we can support another expression: "dimmed" (i.e. "Hide all nodes/edges that are currently dimmed"). Again, not sure how the UI experience would be doing that, but that seems to be the desire.

**Reply by hassanbsee2071** [reply]
For me what could be the possible solution, my end goal is to visualize full traffic flow from start to end and I have a lot many namespaces. How can I visualize easily?

**Reply by jmazzitelli** [reply]
As @jshaughn [mentioned earlier](https://github.com/kiali/kiali/discussions/7892#discussioncomment-11155359), if you have a lot of services and traffic connections and namespaces, it is simply normal to expect a complicated graph. It's "the nature of the beast" when you have a large mesh - you will have to tell Kiali what you are looking for, otherwise, it doesn't know so it will just show you everything. So you will have to figure out the best way in your situation how to minimize/filter the graph for what you are looking for using Kiali UI's features.

To say `visualize full traffic flow from start to end` is not descriptive enough - if you have a lot of requests flowing, the question becomes WHAT traffic flow are you looking for? Are you looking for: A specific request from a specific source? Traffic flowing through a particular service? Full traffic flow that has errors in them? WHAT are you looking for? Without telling Kiali that kind of info, it will just show you everything, which can be a large spider web on large meshes, as you see. 

So you will have to determine what you are trying to see, what you are trying to narrow down, and then based on that, you can add expressions in the hide/find text boxes to help narrow it down, or use the other options such as filtering out different protocols (do you just want to see TCP? HTTP?) and things like that. Play with the Display options and hide/find options in the Kiali UI to see what you can do. As Jay mentioned, read the Help dialogue for the hide/find text fields - there's alot of expressions it supports.

**Reply by hassanbsee2071** [reply]
Thank you @jshaughn . Really appreciate your kind help . 
I got it, I will see how can I narrow down for application. From full traffic flow what I meant is for example take flight-api-v3. Traffic comes from nginx ingress-->kong-->flights-api-v3-->shared-miscellaneous-gateway.shared-->shared-miscellaneous-core.shared-->oms-pricing.oms-->thirdparty service. This is I want to see.

**Reply by jshaughn** [reply]
Just out of curiosity, are you also using tracing?

**Reply by hassanbsee2071** [reply]
So the problem is, different teams handles different namespaces, all developers from different team are not propagating x-request-id. So we are not able to track complete request.
